### PR TITLE
CD-242902 Previously when reading a non-nil, expired token from cache…

### PR DESCRIPTION
**Pull request by johnwu96822**
…, it would not attempt to read from the secondary cache. This is now fixed so that if any read from the primary fails, it will try the secondary cache.

- [ ] @johnny-lai 
- [x] @abdulchaudhrycoupa 
- [ ] @supriya

### CODE REVIEW

File: lib/sand/cache.rb [hunk 1]
```diff
@@ -15,26 +15,61 @@ def initialize(cache:, max_size: MEMORY_MAX_SIZE)
...
+    # This returns the raw data from the cache, which wraps the data with a hash
+    # with keys :data and :expires_at_epoch_time
+    def read_raw(key, margin = 0)
+      hash = @primary.read(key)
+      return read_secondary_raw(key, margin) if hash.nil? || !hash.is_a?(Hash)
+
+      expiry = hash[:expires_at_epoch_time]
+      return read_secondary_raw(key, margin) if expiry.nil?
+
+      if Time.now.to_f >= expiry - margin.to_f
```
Comment (abdulchaudhrycoupa) [inline lib/sand/cache.rb:32]: @johnwu96822 time zone utc ?

Comment (johnwu96822) [reply lib/sand/cache.rb:32]: `Time.now.to_f` (or to_i) returns the epoch time. So no need for utc.

Comment (abdulchaudhrycoupa) [reply lib/sand/cache.rb:32]: 👍🏽

**Review by abdulchaudhrycoupa — APPROVED** [review]

**Review by johnny-lai — APPROVED** [review]

**Review by supriya — APPROVED** [review]

**Comment by johnwu96822** [comment]
Tested by hand. Token is no longer valid by the time you read this.

In this test I got a token, stored in both primary and secondary caches. Then I overwrite the token in the primary so that the data is expired. Then when I made another token call, it saw that the primary's data was expired, so it tried the secondary and found the unexpired data. So it copied this data to the primary and returned.
```
[42] pry(main)> XXX::Sand.client.token
=> "hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI"
[43] pry(main)> XXX::Sand.client.cache.primary.read('xxxxxx/sand/resources')
=> {:data=>"hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI", :expires_at_epoch_time=>1622756074.9935548}
[44] pry(main)> XXX::Sand.client.cache.secondary.read('xxxxxx/sand/resources')
=> {:data=>"hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI", :expires_at_epoch_time=>1622756074.9935548}
[45] pry(main)> XXX::Sand.client.cache.primary.write('xxxxxx/sand/resources', {:data=>"bad already expired", :expires_at_epoch_time=>Time.now.to_f}, expires_in: 3600)
=> true
[46] pry(main)> XXX::Sand.client.cache.primary.read('xxxxxx/sand/resources')
=> {:data=>"bad already expired", :expires_at_epoch_time=>1622753218.387192}
[47] pry(main)> XXX::Sand.client.cache.secondary.read('xxxxxx/sand/resources')
=> {:data=>"hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI", :expires_at_epoch_time=>1622756074.9935548}
[48] pry(main)> XXX::Sand.client.token
=> "hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI"
[49] pry(main)> XXX::Sand.client.cache.primary.read('xxxxxx/sand/resources')
=> {:data=>"hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI", :expires_at_epoch_time=>1622756074.9935548}
[50] pry(main)> XXX::Sand.client.cache.secondary.read('xxxxxx/sand/resources')
=> {:data=>"hHsbWZ2roYdFrEAPagW56RILEBUjpzkOembSyc3lYLA.okIAN672DXmhFsxyMGS_oAyPrp5CeiRlKiNGz8fSdhI", :expires_at_epoch_time=>1622756074.9935548}
```

**Comment by johnwu96822** [comment]
[lockit v4.9.3] merged via Lockit CLI on 2021-06-03 14:20:49.477356 -0700 PDT m=+5.111811179

**Pull request merged**